crates/llm-local/src/token_output_stream.rs

```diff
@@ -0,0 +1,75 @@
+/// This is a wrapper around a tokenizer to ensure that tokens can be returned to the user in a
+/// streaming way rather than having to wait for the full decoding.
+/// Implementation for TokenOutputStream Code is borrowed from
+///
+/// Borrowed from https://github.com/huggingface/candle/blob/main/candle-examples/src/token_output_stream.rs
+/// (Commit SHA 4fd00b890036ef67391a9cc03f896247d0a75711)
+pub struct TokenOutputStream {
+    tokenizer: tokenizers::Tokenizer,
+    tokens: Vec<u32>,
+    prev_index: usize,
+    current_index: usize,
+}
+
+impl TokenOutputStream {
+    pub fn new(tokenizer: tokenizers::Tokenizer) -> Self {
+        Self {
+            tokenizer,
+            tokens: Vec::new(),
+            prev_index: 0,
+            current_index: 0,
+        }
+    }
+
+    /// Processes the next token in the sequence, decodes the current token stream,
+    /// and returns any newly decoded text.
+    ///
+    /// Based on the following code: <https://github.com/huggingface/text-generation-inference/blob/5ba53d44a18983a4de32d122f4cb46f4a17d9ef6/server/text_generation_server/models/model.py#L68>
+    pub fn next_token(&mut self, token: u32) -> anyhow::Result<Option<String>> {
+        let prev_text = if self.tokens.is_empty() {
+            String::new()
+        } else {
+            let tokens = &self.tokens[self.prev_index..self.current_index];
+            self.decode(tokens)?
+        };
+        self.tokens.push(token);
+        let text = self.decode(&self.tokens[self.prev_index..])?;
+        if text.len() > prev_text.len() && text.chars().last().unwrap().is_alphanumeric() {
+            let (_, text) = text.split_at(prev_text.len());
+            self.prev_index = self.current_index;
+            self.current_index = self.tokens.len();
+            Ok(Some(text.to_string()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Decodes the remaining tokens and returns any new text found.
+    ///
+    /// This function decodes tokens from `self.prev_index` to the end and
+    /// compares it with the previously decoded portion (from `self.prev_index`
+    /// to `self.current_index`). If new text is found, it returns the
+    /// additional part as `Some(String)`. Otherwise, returns `None`.
+    pub fn decode_rest(&self) -> anyhow::Result<Option<String>> {
+        let prev_text = if self.tokens.is_empty() {
+            String::new()
+        } else {
+            let tokens = &self.tokens[self.prev_index..self.current_index];
+            self.decode(tokens)?
+        };
+        let text = self.decode(&self.tokens[self.prev_index..])?;
+        if text.len() > prev_text.len() {
```

Comment:
We're decoding the previous text just to check its length against the new text. I still don't fully understand why that's necessary, and it seems fairly wasteful. Can you explain why this check is necessary?

Reply:
That is not a part I fully understand either. It is a lift and shift from the [`candle` samples](https://github.com/huggingface/candle/blob/main/candle-examples/src/token_output_stream.rs)

Reply:
Not sure I follow the part with "just to check". It's also decoding and returning the actual text (conditionally, if there is new text generated).

Reply:
Right, but we're returning part of `text` (not `prev_text`). The only information from `prev_text` we use is it's byte length.

Reply:
Okay I have spent a little more time staring at this, My understanding now is that adding a token does not mean we have a new character and also addition of a token might introduce a non alphanumeric character because we currently are in the middle of a bunch of tokens that make an actual word. We can definitely look at just storing the length of the prev text to avoid the extra `decode`. 

alternatively one other thing we could do is drop the `token_output_stream` on this PR and just decode once the entire content is generated since we do not stream the responses back either way. This approach is strictly no worse than what we have today.

Reply:
YAGNI I'd be in favor of dropping the streaming for now - since we're not actually using it, it's simply more maintenance overhead.

Reply:
I have pushed a commit removing it. If everything looks good, I will squash the commits.